File: memento/typescript/Memento.ts
// Memento
class Memento {
  private content: string;

  constructor(content: string) {
    this.content = content;
  }

  getContent(): string {
    return this.content;
  }
}

// Originator
class TextEditor {
  private content: string = '';

  setContent(content: string): void {
    this.content = content;
  }

  getContent(): string {
    return this.content;
  }

  save(): Memento {
    return new Memento(this.content);
  }

  restore(memento: Memento): void {
    this.content = memento.getContent();
  }
}

// Caretaker
class History {
  private mementos: Memento[] = [];

  push(memento: Memento): void {
    this.mementos.push(memento);
  }

  pop(): Memento | undefined {
    return this.mementos.pop();
  }
}

// Usage
const editor = new TextEditor();
const history = new History();

editor.setContent('Version 1');
history.push(editor.save());
console.log(`Current: ${editor.getContent()}`);

editor.setContent('Version 2');
history.push(editor.save());
console.log(`Current: ${editor.getContent()}`);

editor.setContent('Version 3');
console.log(`Current: ${editor.getContent()}`);

const memento1 = history.pop();
if (memento1) editor.restore(memento1);
console.log(`After undo: ${editor.getContent()}`);

const memento2 = history.pop();
if (memento2) editor.restore(memento2);
console.log(`After undo: ${editor.getContent()}`);

File: memento/java/Memento.java
import java.util.ArrayList;
import java.util.List;

// Memento
class Memento {
    private String content;

    public Memento(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }
}

// Originator
class TextEditor {
    private String content = "";

    public void setContent(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }

    public Memento save() {
        return new Memento(content);
    }

    public void restore(Memento memento) {
        content = memento.getContent();
    }
}

// Caretaker
class History {
    private List<Memento> mementos = new ArrayList<>();

    public void push(Memento memento) {
        mementos.add(memento);
    }

    public Memento pop() {
        if (mementos.isEmpty()) {
            return null;
        }
        return mementos.remove(mementos.size() - 1);
    }
}

// Main class demonstrating Memento pattern
public class Memento {
    public static void main(String[] args) {
        TextEditor editor = new TextEditor();
        History history = new History();

        editor.setContent("Version 1");
        history.push(editor.save());
        System.out.println("Current: " + editor.getContent());

        editor.setContent("Version 2");
        history.push(editor.save());
        System.out.println("Current: " + editor.getContent());

        editor.setContent("Version 3");
        System.out.println("Current: " + editor.getContent());

        editor.restore(history.pop());
        System.out.println("After undo: " + editor.getContent());

        editor.restore(history.pop());
        System.out.println("After undo: " + editor.getContent());
    }
}
